### ISSUE TITLE
[Bug]: /.well-known/webfinger and /.well-known/nodeinfo tests use HTTP and fail

### ISSUE THREAD

**Issue by p4block** [nextcloud/server#41647]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

![image](https://github.com/nextcloud/server/assets/5960575/afbf6336-d9ff-4cc1-8a86-629ab82d34f4)

![image](https://github.com/nextcloud/server/assets/5960575/cac42bbb-cba7-4305-a95a-5279e24c514b)

Despite 99.99% of people mentioning this problem having some kind of incorrect redirect, I think I am _not_ one of them.
The URL its testing is not using HTTPs. **These are the only URLs in the entire Nextcloud instance that get generated wrong**. 

'overwriteprotocol' => 'https' is set and there are no other problems.

### Steps to reproduce

1. Install Nextcloud behind a reverse proxy, I am using Helm and Traefik.
2. Setup redirects properly. I did it in Traefik.
3. Curl the endpoint to see it's properly returning `{"message":"webfinger not supported"}`                                                                                               
4. Frontend test still complains about it, generating the problem in the browser console

### Expected behavior

Test is done over HTTPS and sees that my endpoint actually works

### Installation method

Other Community project

### Nextcloud Server version

27

### Operating system

Other

### PHP engine version

PHP 8.2

### Web server

Nginx

### Database engine version

PostgreSQL

### Is this bug present after an update or on a fresh install?

Upgraded to a MAJOR version (ex. 22 to 23)

### Are you using the Nextcloud Server Encryption module?

Encryption is Disabled

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [X] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "htaccess.RewriteBase": "\/",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "cloud.fosc.space",
            "nextcloud"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "27.1.3.2",
        "overwrite.cli.url": "https:\/\/cloud.fosc.space",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "ldapProviderFactory": "OCA\\User_LDAP\\LDAPProviderFactory",
        "skeletondirectory": "",
        "loglevel": 2,
        "maintenance": false,
        "theme": "",
        "default_phone_region": "ES",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": "6379",
            "password": "***REMOVED SENSITIVE VALUE***"
        },
        "overwriteprotocol": "https",
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "memories.exiftool": "\/var\/www\/html\/custom_apps\/memories\/bin-ext\/exiftool-amd64-glibc",
        "memories.vod.path": "\/var\/www\/html\/custom_apps\/memories\/bin-ext\/go-vod-amd64",
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "mail_smtpsecure": "",
        "mail_smtpauth": false,
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - cloud_federation_api: 1.10.0
  - contactsinteraction: 1.8.0
  - cospend: 1.5.10
  - dav: 1.27.0
  - deck: 1.11.1
  - federatedfilesharing: 1.17.0
  - files: 1.22.0
  - files_pdfviewer: 2.8.0
  - files_reminders: 1.0.0
  - files_rightclick: 1.6.0
  - files_sharing: 1.19.0
  - files_trashbin: 1.17.0
  - files_versions: 1.20.0
  - firstrunwizard: 2.16.0
  - groupfolders: 15.3.1
  - logreader: 2.12.0
  - lookup_server_connector: 1.15.0
  - memories: 6.1.0
  - nextcloud_announcements: 1.16.0
  - notifications: 2.15.0
  - oauth2: 1.15.1
  - photos: 2.3.0
  - privacy: 1.11.0
  - provisioning_api: 1.17.0
  - related_resources: 1.2.0
  - serverinfo: 1.17.0
  - settings: 1.9.0
  - sharebymail: 1.17.0
  - support: 1.10.0
  - survey_client: 1.15.0
  - text: 3.8.0
  - theming: 2.2.0
  - twofactor_backupcodes: 1.16.0
  - updatenotification: 1.17.0
  - user_ldap: 1.17.0
  - user_saml: 5.2.4
  - user_status: 1.7.0
  - viewer: 2.1.0
  - weather_status: 1.7.0
  - workflowengine: 2.9.0
Disabled:
  - activity: 2.19.0 (installed 2.18.0)
  - admin_audit: 1.17.0
  - bruteforcesettings: 2.7.0
  - circles: 27.0.1 (installed 26.0.0)
  - comments: 1.17.0 (installed 1.16.0)
  - dashboard: 7.7.0 (installed 7.3.0)
  - encryption: 2.15.0
  - federation: 1.17.0 (installed 1.16.0)
  - files_external: 1.19.0
  - password_policy: 1.17.0 (installed 1.16.0)
  - recommendations: 1.6.0 (installed 1.5.0)
  - social: 0.6.1 (installed 0.6.1)
  - suspicious_login: 5.0.0
  - systemtags: 1.17.0 (installed 1.16.0)
  - twofactor_totp: 9.0.0
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Nextcloud Logs

```shell
No logs related to this
```

**Comment by kesselb** [comment]
Thank you for taking the time to report a problem :+1: 

The redirect check is done by your browser and therefore uses the same protocol and host. 

You can use the browser network inspector to debug the request/redirects or curl.


```
curl -i "https://cloud.fosc.space/.well-known/webfinger"
HTTP/2 301 
content-type: text/html
date: Tue, 21 Nov 2023 20:29:50 GMT
location: http://cloud.fosc.space/index.php/.well-known/webfinger
referrer-policy: no-referrer
server: nginx/1.25.0
strict-transport-security: max-age=15552000; includeSubDomains; preload
x-content-type-options: nosniff
x-download-options: noopen
x-frame-options: SAMEORIGIN
x-permitted-cross-domain-policies: none
x-robots-tag: noindex, nofollow
x-xss-protection: 1; mode=block
content-length: 169

<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>nginx/1.25.0</center>
</body>
</html>

```

That is not correct. 

As this seems to be a setup issue, I would like to ask you to raise your question at https://help.nextcloud.com

**Comment by p4block** [comment]
@kesselb 

You're right my reverse proxy wasn't working/picking the webfinger regex up, card/webdav was working weirdly though.

However because the entire `/.well-known` endpoint behaved like this, returning http, I just removed every reverse proxy regex for testing and then observed what's happening below.

Nextcloud pod is returning those 301s to "the proper URL but it's http" itself.
This can be seen in the helmchart's nginx config.
https://github.com/nextcloud/helm/blob/4cb7445cb9986add4479f4f97c85e508a0094943/charts/nextcloud/templates/nginx-config.yaml#L105

So in theory a person using the helmchart+nginx doesn't need to configure these redirects themselves.

For example, test:
```
 curl -i https://cloud.fosc.space/.well-known/test  
HTTP/2 301 
content-type: text/html
date: Tue, 21 Nov 2023 23:00:24 GMT
location: http://cloud.fosc.space/index.php/.well-known/test
```

Nginx is not be supported, but it does have the attempt at a production config builtin in the chart. From its perspective it's getting accessed via http sure, but usually it's configured to look for the X-Forwarded-Proto and behave correctly. 

So I think I found a misconfig in the Helmchart's nginx.

Usually you would do this:

`fastcgi_param HTTPS $http_x_forwarded_proto if_not_empty;`

but we have instead a commented out https force and no easy way I can see to configure this.

`#fastcgi_param HTTPS on;` 

So to summarize, I have no clue why my reverse proxy wasn't regexing a raw /.well-known/webfinger, other similar rules I run are working, but that's not the problem here. I would say the unsupported nginx config doesn't work 100% correctly behind the proxy... and I only noticed here where my external redirect regex shenanigans weren't working.

**Comment by p4block** [comment]
Shortly after spending 1 hr writing and rewriting the post I also found why my regex for `https://cloud.fosc.space/.well-known/(nodeinfo|webfinger)` wasn't working and why I ended up in this whole mess. Loving being part of the 99%

Fixing it to actually work caused a redirect loop (oh well) which I haphazardly avoided by matching our entire domain without index.php.

In any case, if the redirect regex regex was working first try I wouldn't have discovered the problem in the nginx config, so something of value was gained. I think. 

Entire Ingress and middleware config ends up like this

```
apiVersion: traefik.io/v1alpha1
kind: IngressRoute
metadata:
  name: nextcloud 
spec:
  entryPoints:
    - websecure
  routes:
    - match: Host(`cloud.fosc.space`)
      kind: Rule
      services:
        - name: nextcloud 
          port: 8080
      middlewares:
        - name: security-headers
        - name: redirect-dav
        - name: redirect-webfinger

---
apiVersion: traefik.io/v1alpha1
kind: Middleware
metadata:
 name: security-headers
spec:
 headers:
   stsSeconds: 15552000
   stsIncludeSubdomains: true
   stsPreload: true


---
apiVersion: traefik.io/v1alpha1
kind: Middleware
metadata:
  name: redirect-dav
spec:
  redirectRegex:
    regex: ^https://(.*)/.well-known/(?:card|cal)dav
    replacement: https://${1}/remote.php/dav
    permanent: true

---
apiVersion: traefik.io/v1alpha1
kind: Middleware
metadata:
  name: redirect-webfinger
spec:
  redirectRegex:
    regex: ^https://cloud.fosc.space/.well-known/(nodeinfo|webfinger)
    replacement: https://cloud.fosc.space/index.php/.well-known/${2}
    permanent: true



```
